src/pykeen/triples/splitting.py

```diff
@@ -0,0 +1,314 @@
+    ratios = _cleanup_ratios(ratios)
+    sizes = _get_group_sizes(triples, ratios)
+
+    if method == 'old':
+        triples_groups = _split_triples(
+            triples,
+            sizes=sizes,
+            random_state=random_state,
+        )
+        # Make sure that the first element has all the right stuff in it
+        logger.debug('cleaning up groups')
+        triples_groups = _tf_cleanup_all(triples_groups, random_state=random_state if randomize_cleanup else None)
+        logger.debug('done cleaning up groups')
+    elif method == 'new' or method is None:
+        triples_groups = _split_triples_with_train_coverage(
+            triples=triples,
+            sizes=sizes,
+        )
+    else:
+        raise ValueError(f'invalid method: {method}')
+
+    for i, (triples, exp_size, exp_ratio) in enumerate(zip(triples_groups, sizes, ratios)):
+        actual_size = triples.shape[0]
+        actual_ratio = actual_size / exp_size * exp_ratio
+        if actual_size != exp_size:
+            logger.warning(
+                f'Requested ratio[{i}]={exp_ratio:.3f} (equal to size {exp_size}), but got {actual_ratio:.3f} '
+                f'(equal to size {actual_size}) to ensure that all entities/relations occur in train.',
+            )
+
+    return triples_groups
+
+
+def _split_triples_with_train_coverage(
+    triples: np.ndarray,
+    sizes: Sequence[int],
+) -> Sequence[np.ndarray]:
+    """
+    Split triples into groups ensuring that all entities and relations occur in the first group of triples.
+
+    :param triples: shape: (num_triples, 3)
+        The triples.
+    :param sizes:
+        The group sizes.
+
+    :return:
+        The groups, where the first group is guaranteed to contain each entity and relation at least once.
+    """
+    seed_mask = _get_cover_randomized_greedy(triples)
+    train_seed = triples[seed_mask]
+    remaining_triples = triples[~seed_mask]
+    # TODO: what to do if train_seed.shape[0] > sizes[0]
+    remaining_sizes = (sizes[0] - train_seed.shape[0],) + tuple(sizes[1:])
+    train, *rest = _split_triples(remaining_triples, remaining_sizes)
+    return (np.concatenate([train_seed, train]), *rest)
+
+
+def _get_cover_deterministic(triples: np.ndarray) -> np.ndarray:
+    """
+    Get a coverage mask for all entities and relations.
+
+    :param triples: shape: (n, 3)
+        The triples.
+
+    :return: shape: (n,)
+        A boolean mask indicating whether the triple is part of the cover.
+    """
+    num_entities = triples[:, [0, 2]].max() + 1
+    num_relations = triples[:, 1].max() + 1
+    num_triples = triples.shape[0]
+
+    # index
+    entities = numpy.full(shape=(num_entities,), fill_value=-1, dtype=numpy.int64)
+    relations = numpy.full(shape=(num_relations,), fill_value=-1, dtype=numpy.int64)
+    h, r, t = triples.T
+    triple_id = numpy.arange(num_triples)
+    entities[h] = relations[r] = entities[t] = triple_id
+
+    if entities.min() < 0:
```

Comment:
currently seems to be failing for some generated triples sets

Reply:
Yes, this is due to some entities not occurring at all in the triples. For "real" triples this should not happen (otherwise the mapping would not be "compact").

Reply:
maybe we should append the generate_triples function to also include a cleanup for entities that don't do this, then a compacting of the identifiers

Reply:
Hopefully solved in d414960e. I wrote tests that check that this is true